system/Entity.php

```diff
@@ -527,69 +474,74 @@ protected function mutateDate($value)
 	 * Provides the ability to cast an item as a specific data type.
 	 * Add ? at the beginning of $type  (i.e. ?string) to get NULL instead of casting $value if $value === null
 	 *
-	 * @param mixed  $value
-	 * @param string $type
+	 * @param mixed  $value     Attribute value
+	 * @param string $attribute Attribute name
+	 * @param string $method    Allowed to "get" and "set"
 	 *
 	 * @return mixed
 	 * @throws Exception
 	 */
-	protected function castAs($value, string $type)
+	protected function castAs($value, string $attribute, string $method)
 	{
+		// Do we need to mutate this into a date?
+		// I don't understand what the dates property is for if it is processed by the same method as datetime!
+		if (in_array($attribute, $this->dates, true))
+		{
+			return $this->mutateDate($value);
+		}
+
+		if (empty($this->casts[$attribute]) || ($method === 'get' && ! $this->_cast))
+		{
+			return $value;
+		}
+
+		$type = $this->casts[$attribute];
+
 		if (strpos($type, '?') === 0)
 		{
 			if ($value === null)
 			{
 				return null;
 			}
+
 			$type = substr($type, 1);
 		}
 
-		switch($type)
+		//In order not to create a separate handler for the json-array type, we transform the required one.
+		$type = $type === 'json-array' ? 'json[array]' : $type;
+
+		if (! in_array($method, ['get', 'set'], true))
 		{
-			case 'int':
-			case 'integer': // alias for 'integer'
-				$value = (int) $value;
-				break;
-			case 'float':
-				$value = (float) $value;
-				break;
-			case 'double':
-				$value = (double) $value;
-				break;
-			case 'string':
-				$value = (string) $value;
-				break;
-			case 'bool':
-			case 'boolean': // alias for 'boolean'
-				$value = (bool) $value;
-				break;
-			case 'csv':
-				$value = explode(',', $value);
-				break;
-			case 'object':
-				$value = (object) $value;
-				break;
-			case 'array':
-				if (is_string($value) && (strpos($value, 'a:') === 0 || strpos($value, 's:') === 0))
-				{
-					$value = unserialize($value);
-				}
+			throw CastException::invalidCastMethod();
+		}
 
-				$value = (array) $value;
-				break;
-			case 'json':
-				$value = $this->castAsJson($value);
-				break;
-			case 'json-array':
-				$value = $this->castAsJson($value, true);
-				break;
-			case 'datetime':
-				return $this->mutateDate($value);
-			case 'timestamp':
-				return strtotime($value);
+		$params = [];
+
+		//Attempt to retrieve additional parameters if specified
+		// type[param, param2,param3]
+		if (preg_match('/^(.+)\[(.+)\]$/', $type, $matches))
+		{
+			$type   = $matches[1];
+			$params = array_map(function ($param) {
+				return trim($param);
+			}, explode(',', $matches[2]));
 		}
 
-		return $value;
+		$type = trim($type, '[]');
+
+		$handlers = array_merge($this->castHandlers, $this->defaultCastHandlers);
```

Comment:
Should be inverted so that castHandles can override defaultCastHandlers

Reply:
I had doubts about replacing the default handlers.
Your opinion helped me decide